### PR TITLE
Refactor `getShardIDs`

### PR DESCRIPTION
**Pull request by wgerges-discovery**

### CODE REVIEW

**Comment by wgerges-discovery** [comment]
### Background
Our particular setup is dealing with billions of stream events daily and our individual consumer process count is high. No amount of jitter is going to keep my team from throttling hard on this call and I suspect we're not in a unique position.

Our dev team is looking at patching this internally. Having looked at the code, I can see what information is required but I can't get a sense for the cadence/frequency that each piece (stream status, shard list, etc.) is required. The ListShards api has a high rate limit, but even the `DescribeStreamSummary` api is still capped at `20 per second`. So, it may be that this library would be better served requesting each separately at different frequencies if possible.


### Problem
 Due to our high volume we started seeing in Kinesis stream in the AWS Console the below error:
`Rate exceeded for stream ... under account ...`
Which limits our ability to view stream metrics and other information.

### Reported issues:
https://github.com/vmware/vmware-go-kcl/issues/61
https://github.com/vmware/vmware-go-kcl/issues/69


### Troubleshooting
AWS support indicated that the error is stemming from calls to the DescribeStream API, which has a limit of 10 requests per second per account. The source of the DescribeStream calls appears to be from the vmware-go-kcl library: (https://github.com/vmware/vmware-go-kcl/blob/ddabb2479438232c72face401551cca99a4beddd/clientlibrary/worker/worker.go#L341), which is used by each sync.

### Solution
AWS support suggested the following: 
“The workaround for this is to use the `ListShards` and `DescribeStreamSummary` APIs together which have higher limits. The combination of these two APIs will give all the information given by the DescribeStream API.”
Note that the `ListShards` API has a limit of 100 requests per second per account, while the `DescribeStreamSummary` API has a limit of 20 requests per second per account; the potential for rate limiting is still present even with the API changes. Additional changes beyond that may be necessary to prevent rate limiting.

File: clientlibrary/worker/worker.go [hunk 1]
```diff
@@ -365,13 +359,12 @@ func (w *Worker) getShardIDs(startShardID string, shardInfo map[string]bool) err
 				EndingSequenceNumber:   aws.StringValue(s.SequenceNumberRange.EndingSequenceNumber),
 			}
 		}
-		lastShardID = *s.ShardId
 	}
 
-	if *streamDesc.StreamDescription.HasMoreShards {
-		err := w.getShardIDs(lastShardID, shardInfo)
+	if listShards.NextToken != nil {
+		err := w.getShardIDs(*listShards.NextToken, shardInfo)
```
Comment (taojwmware) [inline clientlibrary/worker/worker.go]: aws.StringValue

**Comment by wgerges-discovery** [comment]
@taojwmware do you have an idea when this will be in the official release?

**Comment by taojwmware** [comment]
Merged to master.

**Pull request merged**